pkg/bitmap/bitmap.go

```diff
@@ -0,0 +1,212 @@
+		b.bitBlock[blockNum] = newBlock
+		b.length--
+	}
+}
+
+// Clone the Bitmap
+func (b *Bitmap) Clone() *Bitmap {
+	ptr := Bitmap{b.length, make([]uint64, len(b.bitBlock))}
+	copy(ptr.bitBlock, b.bitBlock[:])
+	return &ptr
+}
+
+// ReplaceWith replace the content of b with that of newBitmap
+func (b *Bitmap) ReplaceWith(newBitmap *Bitmap) {
+	b.length = newBitmap.length
+	b.bitBlock = make([]uint64, len(newBitmap.bitBlock))
+	copy(b.bitBlock, newBitmap.bitBlock[:])
+}
+
+// countLengthForBlocks count all 1 bits within bitblock of begin and that of end
+func countLengthForBlocks(bitBlock []uint64, begin int, end int) uint64 {
+	length := uint64(0)
+	if begin > end {
+		return length
+	}
+	beginBlock := begin / 64
+	endBlock := end / 64
+	for i := beginBlock; i <= endBlock; i++ {
+		length += uint64(bits.OnesCount64(bitBlock[i]))
+	}
+	return length
+}
+
+// flipBitmapRange flip the bits within range (begin and end)
+func flipBitmapRange(bitBlock []uint64, begin int, end int) {
+	if begin > end {
+		return
+	}
+	if begin == end {
+		blockNum, mask := begin/64, uint64(1)<<(begin%64)
+		bitBlock[blockNum] = bitBlock[blockNum] ^ mask
+		return
+	}
+	beginBlock := begin / 64
+	endBlock := end / 64
+	if beginBlock == endBlock {
+		bitBlock[endBlock] ^= ((^uint64(0) << uint(begin%64)) & ((uint64(1) << (uint(end) % 64 + 1))-1))
+	} else {
+		bitBlock[beginBlock] ^= ^(^uint64(0) << uint(begin%64))
+		for i := beginBlock; i < endBlock; i++ {
+			bitBlock[i] = ^bitBlock[i]
+		}
+		bitBlock[endBlock] ^= ((uint64(1) << (uint(end) % 64 + 1))-1)
+	}
+}
+
+// removeBitmapRange remove the bits within range (begin and end)
+func removeBitmapRange(bitBlock []uint64, begin int, end int) {
+	if begin > end {
+		return
+	}
+	if begin == end {
+		blockNum, mask := begin/64, uint64(1)<<(begin%64)
+		bitBlock[blockNum] = bitBlock[blockNum] &^ mask
+		return
+	}
+	beginBlock := begin / 64
+	endBlock := end / 64
+	if beginBlock == endBlock {
+		bitBlock[beginBlock] &= (((uint64(1) << uint(begin%64)) - 1) | ^((uint64(1) << (uint(end) % 64 + 1)) - 1))
+	} else {
+		bitBlock[beginBlock] &= ((uint64(1) << uint(begin%64)) - 1)
+		for i := beginBlock + 1; i < endBlock; i++ {
+			bitBlock[i] &= ^bitBlock[i]
+		}
+		bitBlock[endBlock] &= ^((uint64(1) << (uint(end) % 64 + 1)) - 1)
+	}
+}
+
+// RemoveRange remove bits within range (begin and end) for the Bitmap
+func (b *Bitmap) RemoveRange(begin int, end int) {
+	oldRangeLength := countLengthForBlocks(b.bitBlock, begin, end)
+	removeBitmapRange(b.bitBlock, begin, end)
+	newRangeLength := countLengthForBlocks(b.bitBlock, begin, end)
+	b.length += int(newRangeLength - oldRangeLength)
+}
+
+// Flip flip bits within range (begin and end) for the Bitmap
+func (b *Bitmap) Flip(begin int, end int) {
+	oldRangeLength := countLengthForBlocks(b.bitBlock, begin, end)
+	flipBitmapRange(b.bitBlock, begin, end)
+	newRangeLength := countLengthForBlocks(b.bitBlock, begin, end)
+	b.length += int(newRangeLength - oldRangeLength)
```

Comment:
IMO this is a micro optimization that can hurt performance sometimes (and also introduces complexity).

You can alternatively do:
```
removeBitmapRange(b.bitBlock, begin, end)
b.length = countLengthForEntireBitMap()
```

If the range passed in is very large (close to size of bit map), then with your implementation we will have to go through all the blocks twice. But if the range is small, then your implementation is better. What do you think would be the common case?

Reply:
Thanks for your suggestion. The range is changeable, usually, the range is within 0 to MaxUint16 (65535). I will do a benchmark on this.

Reply:
I did some benchmarks on this. It does have some performance improvement when flipping the whole bitmap, but if only flip half of the bitmap, the original methods have better performance.
The BenchmarkFlipNew would use `countLengthForEntireBitmap()` while BenchmarkFlipOld keeps using the original Flip method. Benchmark name ends up with 32766 means flipping half of the bitmap, 65535 means flipping all the bitmap.
The following is the benchmark for them.
I will modify the method based on the benchmark result.

```
goos: linux
goarch: amd64
pkg: gvisor.dev/gvisor/pkg/bitmap
cpu: Intel(R) Xeon(R) CPU E5-2620 v4 @ 2.10GHz
BenchmarkFlipNew32766-32        1000000000               0.0000160 ns/op
BenchmarkFlipNew32766-32        1000000000               0.0000136 ns/op
BenchmarkFlipNew32766-32        1000000000               0.0000088 ns/op
BenchmarkFlipNew32766-32        1000000000               0.0000086 ns/op
BenchmarkFlipNew32766-32        1000000000               0.0000062 ns/op
BenchmarkFlipOld32766-32        1000000000               0.0000078 ns/op
BenchmarkFlipOld32766-32        1000000000               0.0000059 ns/op
BenchmarkFlipOld32766-32        1000000000               0.0000051 ns/op
BenchmarkFlipOld32766-32        1000000000               0.0000059 ns/op
BenchmarkFlipOld32766-32        1000000000               0.0000060 ns/op
BenchmarkFlipNew65535-32        1000000000               0.0000094 ns/op
BenchmarkFlipNew65535-32        1000000000               0.0000053 ns/op
BenchmarkFlipNew65535-32        1000000000               0.0000077 ns/op
BenchmarkFlipNew65535-32        1000000000               0.0000067 ns/op
BenchmarkFlipNew65535-32        1000000000               0.0000085 ns/op
BenchmarkFlipOld65535-32        1000000000               0.0000090 ns/op
BenchmarkFlipOld65535-32        1000000000               0.0000074 ns/op
BenchmarkFlipOld65535-32        1000000000               0.0000113 ns/op
BenchmarkFlipOld65535-32        1000000000               0.0000128 ns/op
BenchmarkFlipOld65535-32        1000000000               0.0000212 ns/op
PASS
```

Reply:
Right. I see that which method performs better depends on the use case. If the range is close to the bitmap size, then the New method works better. Otherwise the old method.

I have not looked too closely at how `BitMap.Flip()` is used by the FDTable. But we should make a decision based on what is the common case there. How does the FDTable flip the BitMap? Does it flip the whole thing or just tiny ranges?

Reply:
> Right. I see that which method performs better depends on the use case. If the range is close to the bitmap size, then the New method works better. Otherwise the old method.
> 
> I have not looked too closely at how `BitMap.Flip()` is used by the FDTable. But we should make a decision based on what is the common case there. How does the FDTable flip the BitMap? Does it flip the whole thing or just tiny ranges?

Currently, `Bitmap.Flip()` is more common to flip the whole bitmap, but the method `Bitmap.RemoveRange()` also faces the same situation, which can remove random range of bitmap so I modified the method as the following code.

```
// RemoveRange remove bits within range (begin and end) for the Bitmap. begin and end are inclusive.
func (b *Bitmap) RemoveRange(begin, end int) {
        blockRange := end/64 - begin/64
        // when the number of removed blocks is larger than half of the length of b.bitBlock
        // we get the length by countLengthForAllBlocks for better performance
        if blockRange > len(b.bitBlock)/2 {
                removeBitmapRange(b.bitBlock, begin, end)
                b.length = int(countLengthForAllBlocks(b.bitBlock))
        } else {
                oldRangeLength := countLengthForBlocks(b.bitBlock, begin, end)
                removeBitmapRange(b.bitBlock, begin, end)
                newRangeLength := countLengthForBlocks(b.bitBlock, begin, end)
                b.length += int(newRangeLength - oldRangeLength)
        }
}

// Flip flip bits within range (begin and end) for the Bitmap. begin and end are inclusive.
func (b *Bitmap) Flip(begin, end int) {
        blockRange := end/64 - begin/64
        // when the number of flipped blocks is larger than half of the length of b.bitBlock
        // we get the length by countLengthForAllBlocks for better performance
        if blockRange > len(b.bitBlock)/2 {
                flipBitmapRange(b.bitBlock, begin, end)
                b.length = int(countLengthForAllBlocks(b.bitBlock))
        } else {
                oldRangeLength := countLengthForBlocks(b.bitBlock, begin, end)
                flipBitmapRange(b.bitBlock, begin, end)
                newRangeLength := countLengthForBlocks(b.bitBlock, begin, end)
                b.length += int(newRangeLength - oldRangeLength)
        }
}
```

Reply:
Whoops I was trying to write pseudo code above in my suggestion. `countLengthForAllBlocks ` should definitely be a method **on** `BitMap`. And so should `countLengthForBlocks`. They should not be standalone functions.

Easier to call them `b. countLengthForBlocks(begin, end)` than `countLengthForBlocks(b.bitBlock, begin, end)`.

Secondly, I think it would be worth having these in separate methods: `BitMap.Flip()` vs `BitMap.FlipRange` and `BitMap.Clear()` vs `BitMap.ClearRange()`. And let the user of `BitMap` decide which one to use. Adding such heuristics which decide which method to use can add complexity. Let the FD table use `BitMap.Flip()` in the code paths in which it intends to flip the whole bitmap. What do you think?

Reply:
Sorry about the mistake, Bitmap.Flip() is **not common** to flip the whole bitmap. It depends on the maximum bit in the bitmaps, like the following code. Only if the maximum equal to the `len(b.bitBlock)*64`, the whole bitmap would get flipped.
```
max = f.fdBitmap.Maximum()
flipFdBitmap.Flip(0, int(max))
```

Reply:
> Secondly, I think it would be worth having these in separate methods: `BitMap.Flip()` vs `BitMap.FlipRange` and `BitMap.Clear()` vs `BitMap.ClearRange()`. And let the user of `BitMap` decide which one to use. Adding such heuristics which decide which method to use can add complexity. Let the FD table use `BitMap.Flip()` in the code paths in which it intends to flip the whole bitmap. What do you think?

As I only write bitmap functions that are used by fd_table. `BitMap.Clear()` is an important method for bitmap, but it is not used by fd_table. From my point of view, it would be better not to contain these methods for now. How do you think?

Reply:
Sounds good, thanks for the explanation. On second thought, I would suggest the following:

- Rename functions to `FlipRange` and `ClearRange`.
- Implement the optimization you described [above](https://github.com/google/gvisor/pull/6208#discussion_r656780401). I think it would be fine, and does not add too much complexity. Just make the comment clear that if the range is greater than half then we do lesser work by just counting 1s for the entire bitmap.